src/charm.py

```diff
@@ -132,15 +144,43 @@ def _on_mongodb_storage_detaching(self, event: ops.charm.StorageDetachingEvent)
         except PyMongoError as e:
             logger.error("Failed to remove %s from replica set, error=%r", self.unit.name, e)
 
+    def update_app_relation_data(self) -> None:
+        """Helper function to update application relation data."""
+        if "db_initialised" not in self.app_data:
+            return
+
+        database_users = set()
+
+        with MongoDBConnection(self.mongodb_config) as mongo:
+            database_users = mongo.get_users()
+
+        for relation in self.model.relations[REL_NAME]:
+            username = self.client_relations._get_username_from_relation_id(relation.id)
+            password = relation.data[self.app]["password"]
+            config = self.client_relations._get_config(username, password)
+            if username in database_users:
+                data = relation.data[self.app]
+                data["endpoints"] = ",".join(config.hosts)
+                data["uris"] = config.uri
+                relation.data[self.app].update(data)
+
     def _relation_departed(self, event: ops.charm.RelationDepartedEvent) -> None:
         """Remove peer from replica set if it wasn't able to remove itself.
 
         Args:
             event: The triggering relation departed event.
         """
-        # allow leader to update hosts if it isn't leaving
-        if self.unit.is_leader() and not event.departing_unit == self.unit:
-            self._update_hosts(event)
+        # allow leader to update relation data and hosts if it isn't leaving
+        if not self.unit.is_leader() or event.departing_unit == self.unit:
+            return
+        self._update_hosts(event)
+
+        try:
+            self.update_app_relation_data()
+        except PyMongoError as e:
```

Comment:
I wonder what kind of errors will be catched here, and I defer should be the treatment for all, i.e. couldn't be the case of this is masking a blocking condition?

Reply:
Let me try to find out the exact error so we can be specific about this.

Reply:
Hmm I think this might be a longer conversation, up to this point the charm has been catching all `PyMongoError`'s and deferring when possible. I think the idea is that if a `PyMongoError` occurs two things can happen:

1. We can defer, 
     - Pros: maybe `mongod` needed to do something (ie reconfigure, re-elect a primary) and the defer gives `mongod` time to do whatever operation is required, and when the hook gets fired again it will succeed on the second time.
     - Cons: maybe something more troubling is going on and this defers forever, then the user will have to manually debug the situation. (in this case we have the `logger.debug` message which gives context to the error)
     
2. Don't defer and let it error
     - Pros: we don't infinitely defer
     - Cons: we don't give `mongod` a chance to fix itself and other hooks might be able to fire further breaking the program and the user will still have to manually debug the situation. (it might be easier to manually debug if they see the raised error here)

I think it is best to leave this as it is since option 1 gives us some possibility of resolving the issue. But I am still new to system design and user design of charms. I think it might be useful here to get our local MongoDB/Charm expert @delgod 's opinion on what we should do here.

Reply:
Errors that we catching here are 1) or connection errors, 2) or a no-primary error, 3) or something similar.
MongoDB eventually recovers itself from such errors, so we just need to retry later.
I believe it is best to defer in such scenarios without blocking the whole charm.